docs/design/enhance-pluggable-policy.md

```diff
@@ -0,0 +1,231 @@
+ 1.  Provide a mechanism to discover who can perform a given action on a given resource.  This is useful for answering questions like, "who can create replication controllers in my namespace".
+
+This proposal adds to and extends existing API to so that authorizers may provide the functionality described above.  It does not attempt to describe how the policies themselves can be expressed, that is up the authorization plugins themselves.
+
+
+## Enhancements to existing Authorization interfaces
+The existing Authorization interfaces are described here: [authorization.md](../authorization.md).  A couple additions will allow the development of an Authorizer that matches based on different rules than the existing implementation.
+
+### Request Attributes
+The existing authorizer.Attributes only has 5 attributes (user, groups, isReadOnly, kind, and namespace).  If we add more detailed verbs, content, and resource names, then Authorizer plugins will have the same level of information available to RESTStorage components in order to express more detailed policy.  The replacement excerpt is below.
+
+A request has 7 attributes that can be considered for authorization:
+  - user - the user-string which a user was authenticated as
+  - groups - the groups to which the user belongs
+  - verb - string describing the requestion action. Today we have: get, list, watch, create, update, and delete
+  - what kind of object is being accessed 
+    - applies only to the API endpoints, such as 
+        `/api/v1beta1/pods`.  For miscelaneous endpoints, like `/version`, the kind is the empty string.
+  - the namespace of the object being access, or the empty string if the
+        endpoint does not support namespaced objects.
+  - content - the runtime.Object being submitted. This only has content for creates and updates
+  - resourceName - the name of the resource being gotten (get-ed) or deleted.
+
+### Authorizer Interface
+The existing Authorizer interface is very simple, but there isn't a way to provide details about allows, denies, or failures.  The extended detail is useful for UIs that want to describe why certain actions are disallowed.  Not all Authorizers will want to provide that information, but for those that do having that capability is useful.  In addition, adding a `GetAllowSubjects` method that returns back the users and groups that can perform a particular action makes it possible to answer questions like, "who can see resources in my namespace" (see [ResourceAccessReview](#ResourceAccessReview) further down).
+
+```
+// OLD
+type Authorizer interface {
+  Authorize(a Attributes) error 
+}
+```
+
+```
+// NEW
+type Authorizer interface {
+  Authorize(a Attributes) (allowed bool, reason string, evaluationError error)
+
+  // GetAllowedSubjects takes a set of attributes, ignores any provided users and groups and returns back
```

Comment:
Requiring the authorizer to provide the full list of users/groups seems expensive if this will be used to check a particular user or group's access. Not sure if there should be two methods (`GetAllowedSubjects` which returns a strict subset of the user/groups in the attributes, and `GetAllAllowedSubjects`) or if providing a user and groups in the attributes should limit the returned results.